### PR TITLE
Improvements to keepalive-connection shedding

### PR DESCRIPTION
**Pull request by wjordan**
### Description
As a followup to df72887170c7ef3614c941c9bdefb4a1f3546ebf, this PR is an attempt to refactor the 'keepalive-connection shedding' logic to improve its performance and reliability.

Summary of changes:

1. Instead of closing the socket by breaking out of the `Server#process_client` loop from the 'keepalive == true' branch, move the connection-shedding logic to modify the `res_info[:keep_alive]` variable in `Request#str_headers`. This writes the `Connection: close` header as part of the final response before closing the socket, allowing clients to reopen a new connection more gracefully (for example, preventing 'socket read' errors from being logged in `wrk`).

2. Instead of shedding the keepalive connection after a fixed number of consecutive inline requests, shed the connection whenever the server is 'at capacity' (busy threads >= max), _and_ there is a new pending connection in the listener socket waiting to be accepted.

3. Additionally, push a keepalive connection back to the reactor whenever there's a pending request in the threadpool backlog (but not a new connection pending), to ensure that requests are serviced fairly across all the keepalive connections already accepted in the server.

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [ ] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

File: lib/puma/server.rb [hunk 1]
```diff
@@ -447,23 +448,17 @@ def process_client(client, buffer)
...
-            break if requests >= @max_fast_inline
-
-            check_for_more_data = @status == :run
+            # As an optimization, try to read the next request from the
+            # socket for a short time before returning to the reactor.
+            fast_check = @status == :run
+
+            # Always pass the client back to the reactor after a reasonable
+            # number of inline requests if there are other requests pending.
+            fast_check = false if requests >= @max_fast_inline &&
```
Comment (nateberkopec) [inline lib/puma/server.rb:457]: `fast_check == requests < @max_fast_inline && thread_pool.backlog == 0`

Comment (johnnyshields) [reply lib/puma/server.rb:457]: he's assigning `fast_check = @status == :run` above. In order to preserve meaning, should be:

```
fast_check = @status == :run && (requests < @max_fast_inline && thread_pool.backlog == 0)
```

**Comment by nateberkopec** [comment]
LGTM and fixes what I saw in `wrk` and `hey`. Should release this soon.

**Comment by nateberkopec** [comment]
I'm gonna backport to 4.3 as well

**Pull request merged**